File: redwood-yoga-java/src/main/java/io/github/orioncraftmc/meditate/internal/detail/Values.java
package io.github.orioncraftmc.meditate.internal.detail;

import io.github.orioncraftmc.meditate.internal.YGValue;
import io.github.orioncraftmc.meditate.internal.enums.YGEdge;
import java.util.ArrayList;
import org.jetbrains.annotations.NotNull;


public class Values<T> //Type originates from: Yoga-internal.h
{
    private final ArrayList<CompactValue> values_;


    public Values() {
        values_ = new ArrayList<>();
    }

    public Values(final @NotNull YGValue defaultValue) {
        values_ = new ArrayList<>();
        values_.add(CompactValue.createCompactValue(defaultValue));
    }

    public Values(final @NotNull YGValue defaultValue, int size) {
        values_ = new ArrayList<>(size);
        values_.add(CompactValue.createCompactValue(defaultValue));
    }

    public final @NotNull CompactValue get(Integer i) {
        return CompactValue.createCompactValue(getValue(i).convertToYgValue());
    }

    private CompactValue getValue(Integer i) {
        while (values_.size() < (i + 1)) {
            values_.add(CompactValue.ofUndefined());
        }
        return values_.get(i);
    }

    public final void set(Integer i, CompactValue value) {
        values_.set(i, value);
    }

    public final @NotNull CompactValue getCompactValue(@NotNull YGEdge edge) {
        return getCompactValue(edge.getValue());
    }

    public final @NotNull CompactValue getCompactValue(Integer i) {
        return CompactValue.createCompactValue(getValue(i).convertToYgValue());
    }

    public final YGValue get(int i) {
        return getValue(i).convertToYgValue();
    }


    public final void set(int i, @NotNull YGValue value) {
        values_.set(i, CompactValue.createCompactValue(value));
    }

    public boolean equalsTo(final @NotNull Values<T> other) {
        for (int i = 0; i < values_.size(); ++i) {
            if (getValue(i) != other.getValue(i)) {
                return false;
            }
        }
        return true;
    }
}

File: redwood-yoga/src/commonMain/kotlin/app/cash/redwood/yoga/detail/Values.kt
package app.cash.redwood.yoga.detail

import app.cash.redwood.yoga.YGValue
import app.cash.redwood.yoga.enums.YGEdge

class Values<T> //Type originates from: Yoga-internal.h
{
    private val values_: MutableList<CompactValue>

    constructor() {
        values_ = ArrayList()
    }

    constructor(defaultValue: YGValue) {
        values_ = ArrayList()
        values_.add(CompactValue.createCompactValue(defaultValue))
    }

    constructor(defaultValue: YGValue, size: Int) {
        values_ = ArrayList(size)
        values_.add(CompactValue.createCompactValue(defaultValue))
    }

    private fun getValue(i: Int): CompactValue {
        while (values_.size < i + 1) {
            values_.add(CompactValue.ofUndefined())
        }
        return values_[i]
    }

    operator fun set(i: Int?, value: CompactValue) {
        values_[i!!] = value
    }

    fun getCompactValue(edge: YGEdge): CompactValue {
        return getCompactValue(edge.ordinal)
    }

    fun getCompactValue(i: Int): CompactValue {
        return CompactValue.createCompactValue(getValue(i).convertToYgValue())
    }

    operator fun get(i: Int): YGValue {
        return getValue(i).convertToYgValue()
    }

    operator fun set(i: Int, value: YGValue) {
        values_[i] = CompactValue.createCompactValue(value)
    }

    fun equalsTo(other: Values<T>): Boolean {
        for (i in values_.indices) {
            if (getValue(i) !== other.getValue(i)) {
                return false
            }
        }
        return true
    }
}
